lib/yang.c

```diff
@@ -621,6 +621,29 @@ static void ly_log_cb(LY_LOG_LEVEL level, const char *msg, const char *path)
 		zlog(priority, "libyang: %s", msg);
 }
 
+const char *yang_print_errors(struct ly_ctx *ly_ctx, char *buf, size_t buf_len)
+{
+	struct ly_err_item *ei;
+	const char *path;
+	int slen = 0;
+
+	ei = ly_err_first(ly_ctx);
+	if (!ei)
+		return "";
+
+	slen += snprintf(buf + slen, buf_len - slen, "YANG error(s):\n");
+	for (; ei; ei = ei->next)
+		slen += snprintf(buf + slen, buf_len - slen, " %s\n", ei->msg);
+
+	path = ly_errpath(ly_ctx);
+	if (path)
+		snprintf(buf + slen, buf_len - slen, " YANG path: %s\n", path);
```

Comment:
Can you switch this code to use `strlcat`? As is we may as well be using `sprintf` here since:
>If the output was truncated due to this limit, then the return value is the number of characters (excluding the terminating null byte) which would have been written to the final string if enough space had been available

Reply:
`strlcat` doesn't do formatting like `snprintf` does and I think this is a common idiom:

```c
char *str = buf;
int slen = 0;

slen += snprintf(str + slen, sizeof(str) - slen, fmt, args);
if (foo)
        slen += snprintf(str + slen, sizeof(str) - slen, fmt, args);
```

Are there any idioms that do this with `strlcat`? If there is and it uses more code, I suppose, how could we make it smaller?

I think this is a repeated review comment and I couldn't think of a better solution yet.

---

BTW: here is a way to reproduce the problem:

```c
#include <stdio.h>
​
int
main(int argc, char *argv[])
{
        char buf[32];
        int slen = 0;
​
        slen += snprintf(buf + slen, sizeof(buf) - slen, "%s", "Hello World!\n");
        while (argc--)
                slen += snprintf(buf + slen, sizeof(buf) - slen, "%s",
                                 "Hello World!\n");
​
        printf("buf='%s'\n", buf);
​
        return 0;
}
```

Compile it and call with multiple arguments:
- no argument works
- 1 argument works
- 2 argument crashes
   ```
   ./foo 1 2
   buf='Hello World!
   Hello World!
   Hello'
   *** stack smashing detected ***: <unknown> terminated
   ```

Reply:
@qlyoung yeah using `snprintf()` that way is super convenient but you're right, it's unsafe. I've changed this function to use `strlcpy()` and `strlcat()` only.

@rzalamena for the cases where we need format strings, I think we can `snprintf()` to a local buffer and pass that buffer to `strlcpy()`. This extra step can be a bit annoying but what can we do :/